File: aoc2023/day01/main.go
package main

import (
	"bufio"
	"bytes"
	"os"
	"strconv"
	"strings"
	"unicode"
)

var (
	digits         = map[string]int{}
	wordsAndDigits = map[string]int{
		"three": 3,
		"seven": 7,
		"eight": 8,
		"four":  4,
		"five":  5,
		"nine":  9,
		"one":   1,
		"two":   2,
		"six":   6,
	}
)

/*
* part 1:  54630
* part 2:  54770
 */
func main() {
	input, err := os.ReadFile("input.txt")
	if err != nil {
		panic(err)
	}
	println("part 1: ", solve(input, digits))
	println("part 2: ", solve(input, wordsAndDigits))
}

// solve
func solve(input []byte, table map[string]int) string {
	scanner := bufio.NewScanner(bytes.NewBuffer(input))
	var sum int
	for scanner.Scan() {
		line := scanner.Text()
		if len(line) == 0 {
			continue
		}
		first, last := -1, 0
		for i, v := range line {
			if unicode.IsDigit(v) {
				last = int(v - '0')
				if first == -1 {
					first = last
				}
			}
			for k, w := range table {
				if strings.HasPrefix(line[i:], k) {
					last = w
					if first == -1 {
						first = last
					}
				}
			}
		}
		sum += int(first*10 + last)
	}

	if err := scanner.Err(); err != nil {
		return err.Error()
	}

	return strconv.Itoa(sum)
}

File: aoc2023/day01/main.rs
fn process(l: &str, words: &Vec<&str>) -> u32 {
    let mut last = 0;
    let mut first = 10;
    for (i, c) in l.chars().enumerate() {
        if c.is_digit(10) {
            last = c.to_digit(10).unwrap();
            if first == 10 {
                first = last;
            }
        } else {
            let s: String = l.chars().skip(i).collect();
            for i in 0..words.len() {
                if s.starts_with(words[i]) {
                    last = i as u32 + 1;
                    if first == 10 {
                        first = last;
                    }
                }
            }
        }
    }
    first * 10 + last
}

fn main() {
    let words: Vec<&str> = vec![
        "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
    ];
    let empty: Vec<&str> = vec![];

    let input = include_str!("./input.txt");
    println!(
        "part1: {}",
        input.lines().map(|l| process(l, &empty)).sum::<u32>()
    );
    println!(
        "part2: {}",
        input.lines().map(|l| process(l, &words)).sum::<u32>()
    );
}
